### PR TITLE
[ Misc ] Support Act Order in Compressed Tensors

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
## Summary ##
Add support for compressed-tensors models which have been quantized using activation ordering (group-wise quantization in decreasing order of activation).
* add `actorder` argument to `CompressedTensorsWNA16`
* add `weight_g_idx` layer parameter
### Accuracy ###
Full Precision
```
vllm (pretrained=Qwen/Qwen2-0.5B-Instruct,add_bos_token=True), gen_kwargs: (None), limit: 250.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|?  |0.384|?  |0.0308|
|     |       |strict-match    |     5|exact_match|?  |0.384|?  |0.0308|
```
Group Quantization Only (`ksayers/gwen_group`)
```
vllm (pretrained=kylesayrs/gwen_group,add_bos_token=True), gen_kwargs: (None), limit: 250.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|?  |0.216|?  |0.0261|
|     |       |strict-match    |     5|exact_match|?  |0.196|?  |0.0252|
```
Activation Ordering (`ksayers/gwen_actorder`)
```
vllm (pretrained=ksayers/gwen_actorder,add_bos_token=True), gen_kwargs: (None), limit: 250.0, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|                             
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|                             
|gsm8k|      3|flexible-extract|     5|exact_match|?  |0.248|?  |0.0274|                             
|     |       |strict-match    |     5|exact_match|?  |0.248|?  |0.0274|
```
### Latency Regression ###
```
Namespace(model=‘/home/ksayers/llm-compressor/gwen_actorder/’, speculative_model=None, num_speculative_tokens=None, speculative_draft_tensor_parallel_size=None, tokenizer=None, quantization=None, tensor_parallel_size=1, input_len=32, output_len=128, batch_size=32, n=1, use_beam_search=False, num_iters_warmup=10, num_iters=30, trust_remote_code=False, max_model_len=None, dtype=‘auto’, enforce_eager=False, kv_cache_dtype=‘auto’, quantization_param_path=None, profile=False, profile_result_dir=None, device=‘auto’, block_size=16, enable_chunked_prefill=False, enable_prefix_caching=False, use_v2_block_manager=False, ray_workers_use_nsight=False, download_dir=None, output_json=None, gpu_memory_utilization=0.9, load_format=‘auto’, distributed_executor_backend=None, otlp_traces_endpoint=None)
```
Group Quantization Only
```
Avg latency: 0.8884373404396076 seconds
10% percentile latency: 0.8715801022946834 seconds
25% percentile latency: 0.8739993472117931 seconds
50% percentile latency: 0.876951577141881 seconds
75% percentile latency: 0.8830150356516242 seconds
90% percentile latency: 0.9393035409972071 seconds
99% percentile latency: 0.9404808702412992 seconds
```
Activation Ordering
```
Avg latency: 0.9159474782645702 seconds
10% percentile latency: 0.9001966264098883 seconds
25% percentile latency: 0.9010569080710411 seconds
50% percentile latency: 0.9041027296334505 seconds
75% percentile latency: 0.9064613012596965 seconds
90% percentile latency: 0.9662564094178379 seconds
99% percentile latency: 0.9761117453686893 seconds
```

### CODE REVIEW

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 1]
```diff
@@ -127,9 +147,15 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         layer.workspace = marlin_make_workspace(
             layer.output_size_per_partition, device)
 
-        # Act-order not supported in compressed-tensors yet, so set to empty.
-        layer.g_idx = marlin_make_empty_g_idx(device)
-        layer.g_idx_sort_indices = marlin_make_empty_g_idx(device)
+        # Handle sorting for activation reordering if needed.
+        if self.act_order:
+            breakpoint()
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:152]: this is all -1 right now

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 2]
```diff
@@ -21,6 +22,7 @@ class CompressedTensorsWNA16(CompressedTensorsScheme):
     def __init__(self,
                  strategy: str,
                  num_bits: int,
+                 act_order: bool,
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:25]: Shouldn't act_order have a default value since it isn't in old configs?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 3]
```diff
@@ -21,6 +22,7 @@ class CompressedTensorsWNA16(CompressedTensorsScheme):
...
                  group_size: Optional[int] = None):
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py]: Nit: act_order should be after group_size, since you can't have it without group_size being set

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 4]
```diff
@@ -190,7 +190,9 @@ def _get_scheme_from_parts(
                 return CompressedTensorsWNA16(
                     num_bits=weight_quant.num_bits,
                     strategy=weight_quant.strategy,
-                    group_size=weight_quant.group_size)
+                    act_order=weight_quant.actorder,
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:193]: is it act_order or actorder?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 5]
```diff
@@ -37,6 +39,12 @@ def __init__(self,
         else:
             self.group_size = group_size
 
+        if act_order and self.group_size == -1:
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:42]: Should we check that group_size isn't None if act_order is set?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:193]: its actorder, I updated everything to be act_order tho

**Review by mgoin — APPROVED** [review]
Thanks for changes, LGTM with model smoke test

**Comment by alexm-neuralmagic** [comment]
LGTM

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 6]
```diff
@@ -119,14 +127,21 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
                                                           dtype=torch.int64),
                                          weight_loader=weight_loader)
 
+        # G_IDX (for activation reordering)
+        g_idx = BasevLLMParameter(data=torch.empty(input_size_per_partition,
+                                                   dtype=torch.int32),
+                                  weight_loader=weight_loader)
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:133]: Is it okay to make this parameter in every case? What about older checkpoints that don't have this parameter?

Comment (mgoin) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:133]: I think my question was worded weirdly, sorry. I am just concerned about the weight loader trying to find this parameter in the checkpoint, and it not being present.

Comment (kylesayrs) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:133]: I regression tested using `neuralmagic/TinyLlama-1.1B-Chat-v1.0-marlin` without issue

**Comment by kylesayrs** [comment]
Do not merge, tensor parallel bug needs to be fixed

**Comment by kylesayrs** [comment]
False alarm on tensor parallelism bug. Regression testing was performed with `TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T` with and without activation ordering and `tensor_parallel_size=2`

**Comment by kylesayrs** [comment]
Moving to draft while support for static_grouping actorder is added

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py (hunk 6 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:139]: I'd just update to only create the parameter if `actorder` is True

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 7]
```diff
@@ -119,14 +127,21 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
                                                           dtype=torch.int64),
                                          weight_loader=weight_loader)
 
+        # G_IDX (for activation reordering)
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:136]: Can you add a test case for this case?
1. tests/quantization/test_compressed_tensors.py
2. add a model to models.txt under tests/weight_loading

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py (hunk 6 above)
Comment (kylesayrs) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:139]: This is because the g_idx passed to the kernel is conditional on the `actorder` flag in the config
https://github.com/vllm-project/vllm/pull/6358/files#diff-df5f822218e5ac1430f35a806bc9cebd78c99cfe1e6738de89ae3e9f5a1fdbecR162

**Review by kylesayrs — APPROVED** [review]

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py (hunk 6 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:139]: If self.actorder is True, it'll use the created parameter. Otherwise, it'll create an empty one. So I dont think you need to initialize it here if `self.actorder` is False

Comment (mgoin) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:139]: Yeah that seems to be the case from this else-case later - so no need to make the parameter
```python
            layer.weight_g_idx = marlin_make_empty_g_idx(device)
            layer.g_idx_sort_indices = marlin_make_empty_g_idx(device)
```

**Review by mgoin — CHANGES_REQUESTED** [review]
LGTM once the remaining issues are addressed

**Comment by kylesayrs** [comment]
New requirements have been added to act-order to support different strategies such as `weight-only` ordering and `group` ordering. See https://github.com/neuralmagic/compressed-tensors/pull/146

I've made a PR I'd like to merge into this branch which conditions activation ordering on g-idx directly rather than relying on the config

**Comment by kylesayrs** [comment]
Moved to https://github.com/vllm-project/vllm/pull/8135
